### PR TITLE
Code drop of canton community

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**

### CODE REVIEW

File: ci/drop-canton.sh [hunk 1]
```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env bash
+# Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+
+set -xeuo pipefail
+
+date=$(date -u "+%Y-%m-%d %H:%M")
+canton_dir=canton
+root_private="/run/user/1000/tmp.W76RmKNhce"
```
Comment (rgugliel-da) [inline ci/drop-canton.sh:10]: What is the goal of this directory ? Should you use `root_canton` (temp directory created below) instead ?

Comment (remyhaemmerle-da) [inline ci/drop-canton.sh:10]:
```suggestion
```

File: ci/cron/daily-compat.yml [hunk 2]
```diff
@@ -173,6 +173,49 @@ jobs:
                                        --project daml
           fi
 
+  - job: drop canton
```
Comment (garyverhaegen-da) [inline ci/cron/daily-compat.yml:176]: Aspirational job name. I like it!

File: ci/drop-canton.sh [hunk 3]
```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env bash
+# Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+
+set -xeuo pipefail
```
Comment (garyverhaegen-da) [inline ci/drop-canton.sh:6]: Generally speaking we don't want `-x` on CI as that risks exposing credentials in various ways.

File: ci/drop-canton.sh [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  local -ra cmd=( "$@" )
+  local output
+
+  debug "$message [${cmd[*]}]"
+
+  if ! output=$("$@" 2>&1); then
+    err "$message failed: $output"
+    exit 1
+  fi
+}
```
Comment (garyverhaegen-da) [inline ci/drop-canton.sh:34]: That seems like a lot of complexity for very little benefit over a couple `echo`'s.

File: ci/drop-canton.sh [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+function refresh() {
```
Comment (garyverhaegen-da) [inline ci/drop-canton.sh:36]: I tend to [prefer subshell functions](https://cuddly-octo-palm-tree.com/posts/2021-10-31-better-bash-functions/).

File: ci/drop-canton.sh [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  file=$1
+  src=${root_private}/${file}
+  dst=${canton_dir}/${file}
+  dir_dst=$(dirname $dst)
```
Comment (garyverhaegen-da) [inline ci/drop-canton.sh:40]: I assume you mean all of those to be `local`.

File: ci/drop-canton.sh [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+  if ! output=$("$@" 2>&1); then
+    err "$message failed: $output"
+    exit 1
+  fi
+}
+
+function refresh() {
+  file=$1
+  src=${root_private}/${file}
```
Comment (garyverhaegen-da) [inline ci/drop-canton.sh:38]: What is `root_private` supposed to be here?

File: ci/drop-canton.sh [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  dir_dst=$(dirname $dst)
+  if [ -e ${dst} ]; then
+    run "Removing ${dst} in preparation for refresh" rm -rf ${dst}
+  fi
+  if [ ! -d ${dir_dst} ]; then
+    run "Creating the ${dir_dst} directory" mkdir -p ${dir_dst}
+  fi
+  run "Copying from ${src} to ${dst}" cp -a ${src} ${dst}
+  run "Adding ${dst} to git" git add ${dst}
+}
```
Comment (garyverhaegen-da) [inline ci/drop-canton.sh:49]: Do we really need such a complex function to do `cp -r $src $dest && git add $dest`?

File: ci/drop-canton.sh (hunk 7 above)
Comment (rgugliel-da) [reply ci/drop-canton.sh:38]: Yep, it fails locally: variable is not defined

File: ci/drop-canton.sh [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  if git diff-index --quiet HEAD ${canton_dir}; then
+
+    echo "No code changes - skipping PR"
+
+  else
+
+    message="Update Canton Community source ${date}"
+    ref_commit="Reference commit: ${hash_private}"
+
+    run "Committing ${message} ${ref_commit}" git commit -m "${message}" -m "${ref_commit}"
```
Comment (garyverhaegen-da) [inline ci/drop-canton.sh:64]: How does `git commit` behave with multiple `-m` arguments? I'd have assumed that to be invalid.

File: ci/drop-canton.sh [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+}
+
+root_canton=$(mktemp -d)
+
+trap "rm -rf ${root_canton}" EXIT
+
+run "clone canton private repo" git clone --branch main ${canton_repo} --depth 1 ${root_canton}
+
+hash_private=$(cd $root_canton; git rev-parse HEAD)
```
Comment (garyverhaegen-da) [inline ci/drop-canton.sh:76]: You can save a `cd` with a `-C` if you want. Why `hash_private` and not e.g. `canton_sha`?

File: ci/drop-canton.sh [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+# Canton Community directories and files inherited from private repo
+for artifact in community; do
```
Comment (garyverhaegen-da) [inline ci/drop-canton.sh:79]: Do we need a loop for a single directory?

**Comment by garyverhaegen-da** [comment]
Broader comment: why do we need to try and keep track of the canton sha at all? Why not just something like (untested):
```
tmp=$(mktemp -d)
git clone git@github.com:DACH-NY/canton.git $tmp
rm -rf canton
cp -r $tmp/community/ canton/
git add canton
if git diff --exit-code -- canton; then
  git commit -m "update canton to $(git -C $tmp rev-parse HEAD)"
fi
```

File: ci/drop-canton.sh [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+function refresh() {
+  file=$1
+  src=${root_private}/${file}
+  dst=${canton_dir}/${file}
+  dir_dst=$(dirname $dst)
+  if [ -e ${dst} ]; then
+    run "Removing ${dst} in preparation for refresh" rm -rf ${dst}
+  fi
+  if [ ! -d ${dir_dst} ]; then
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply ci/drop-canton.sh:44]: Shamelessly copy from https://github.com/DACH-NY/canton/blob/8582b4f1f26cac1dae657580288bec2ac3c1e1c3/release/propose-open-source-code-drop.sh#L135.
It is also the first solution google gives to me when I ask "How to test if a directory exists in bash ?".
What can I do that in a better/clearer way ?

File: ci/drop-canton.sh (hunk 8 above)
Comment (remyhaemmerle-da) [reply ci/drop-canton.sh:49]: Here again I copy from the existing code that drop canton. 
I am open to any suggestion to improve it. 
However, your suggestion do not deletes the file that are not in `$src`.

File: ci/drop-canton.sh (hunk 9 above)
Comment (remyhaemmerle-da) [reply ci/drop-canton.sh:64]: from `git commit --help`:
```
 -m <msg>, --message=<msg>
           Use the given <msg> as the commit message. If multiple -m options are given, their values are concatenated as separate paragraphs.
```

File: ci/drop-canton.sh (hunk 10 above)
Comment (remyhaemmerle-da) [reply ci/drop-canton.sh:76]: :+1:

File: ci/drop-canton.sh (hunk 11 above)
Comment (remyhaemmerle-da) [reply ci/drop-canton.sh]: We need to loop on actually two directories (I missed one)

**Comment by remyhaemmerle-da** [comment]
```
tmp=$(mktemp -d)
git clone git@github.com:DACH-NY/canton.git $tmp
rm -rf canton
cp -r $tmp/community/ canton/
git add canton
if git diff --exit-code -- canton; then
  git commit -m "update canton to $(git -C $tmp rev-parse HEAD)"
fi
```

> Broader comment: why do we need to try and keep track of the canton sha at all? Why not just something like (untested):
> 
> ```
> tmp=$(mktemp -d)
> git clone git@github.com:DACH-NY/canton.git $tmp
> rm -rf canton
> cp -r $tmp/community/ canton/
> git add canton
> if git diff --exit-code -- canton; then
>   git commit -m "update canton to $(git -C $tmp rev-parse HEAD)"
> fi
> ```

I do not want

File: ci/cron/daily-compat.yml [hunk 13]
```diff
@@ -173,6 +173,60 @@ jobs:
...
+        for path in community daml-common-staging; do
+          src=$tmp/$path
+          dst=canton/$path
+          rm -rf $dst
+          mkdir -p $(dirname $dst)
+          cp -rf $src $dst
+          git add $dst
+        done
+
+        canton_sha=$(git -C $tmp rev-parse HEAD)"
```
Comment (remyhaemmerle-da) [inline ci/cron/daily-compat.yml:210]: I need the sha for the name of the branch and the commit message

File: ci/cron/daily-compat.yml [hunk 14]
```diff
@@ -173,6 +173,60 @@ jobs:
...
+
+        git fetch
+        git checkout origin/main
+
+        tmp=$(mktemp -d)
+        trap "rm -rf ${tmp}" EXIT
+
+        git clone git@github.com:DACH-NY/canton.git --depth 1  $tmp
+
+        for path in community daml-common-staging; do
```
Comment (remyhaemmerle-da) [inline ci/cron/daily-compat.yml:201]: I would like to keep this a bit generic.
`path` can be a directory or a file and can be arbitrary nested in the directory structure of canton.

File: ci/cron/daily-compat.yml [hunk 15]
```diff
@@ -173,6 +173,60 @@ jobs:
...
+          src=$tmp/$path
+          dst=canton/$path
+          rm -rf $dst
```
Comment (remyhaemmerle-da) [inline ci/cron/daily-compat.yml:204]: I cannot delete the whole `canton` directory as I will need to put some code here (e.g. bazel rules) that does not exist on the canton side.

File: ci/cron/daily-compat.yml [hunk 16]
```diff
@@ -173,6 +173,60 @@ jobs:
...
+        done
+
+        canton_sha=$(git -C $tmp rev-parse HEAD)"
+        branch="canton-update-${canton_sha}"
+
+        if git diff --exit-code -- canton; then
+            echo "Already up-to-date with latest Canton source."
+        else
+            if [ "main" = "$(Build.SourceBranchName)" ]; then
+                open_pr "$branch" "update canton to $canton_sha"
```
Comment (garyverhaegen-da) [inline ci/cron/daily-compat.yml:217]:
```suggestion
                open_pr "$branch" "update canton to ${canton_sha:0:8}"
```

**Review by garyverhaegen-da — APPROVED** [review]
LGTM.

**Pull request merged**